### PR TITLE
Fix TSV metadata indexation and remove unused lines from bids_dataframe based on split_method

### PR DESCRIPTION
**Pull request by mariehbourget**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR is in complement to PR #1103 after feedback from a user in #1096 for training with microscopy data.

What is done:
1. Fix bug in TSV metadata indexation in `bids_dataframe`:

    -  Previously, the `participant_id` and `sample_id` were assigned to each file based on `subject` and `sample` columns, and then used to merge the TSV metadata. However, `participant_id` and `sample_id` are attributes from the TSV files and should not be pre-assigned.
    - In theory, all `subjects` should have a `participant_id` row in `participants.tsv`, same for `samples` and `sample_id` from `samples.tsv`. However, in practice, it is not always the case and the BIDS-validator doesn't check for this.
    - This was problematic because we often use `participant_id` to split the dataset (see [split_method](https://ivadomed.org/configuration_file.html#split-method)), but we were not taking potential discrepancies between `subjects` and `participant_id` into account.
    - The fix doesn't take retrocompatibility into account as it was more a bug then a feature.
2.  Remove unused files in the `bids_dataframe` based on the `split_method` for training and testing.

    - Related in part to point 1, `split_method` accepts ANY metadata present in the dataframe to make the split. Files that don't have a value in this metadata column won't be used, but they were still present in the `bids_dataframe.csv` file.
    - They are now removed from the dataframe and give a warning that the file is skipped in the terminal, similar to the warning used when derivatives are missing for a subject.
    - This does not apply with the command "segment" as all the files are segmented regardless of `split_method`.

3. Minor related clarification to documentation.

## Linked issues
Related to #1096

### CODE REVIEW

File: ivadomed/loader/bids_dataframe.py [hunk 1]
```diff
@@ -174,23 +190,30 @@ def add_tsv_metadata(self, df, path_data, layout):
...
         if fname_samples.exists():
             df_samples = pd.read_csv(str(fname_samples), sep='\t')
-            df = pd.merge(df, df_samples, on=['participant_id', 'sample_id'], suffixes=("_x", None),
-                          how='left')
+            df_samples['sample'] = df_samples['sample_id'].str.split("sample-").apply(lambda x: x[1])
+            df_samples['subject'] = df_samples['participant_id'].str.split("sub-").apply(lambda x: x[1])
+            columns = df_samples.columns.tolist()
+            if 'participant_id' in df.columns:
+                columns.remove('participant_id')
+            df = pd.merge(df, df_samples[columns], on=['subject', 'sample'], suffixes=("_x", None), how='left')
```
Comment (dyt811) [inline ivadomed/loader/bids_dataframe.py:216]: Need to test this more interactively. To be updated post follow-up.

**Review by dyt811 — COMMENTED** [review]
Thanks, Marie-Hélène! Overall looks great. I do need to test the add_tsv_metadata part a bit more closely so will step through an appropriate example post follow up with you and will update review accordingly.

File: ivadomed/loader/bids_dataframe.py (hunk 1 above)
Comment (dyt811) [reply ivadomed/loader/bids_dataframe.py:216]: @mariehbourget I have stepped through these code as  they are currently implemented, while I was able to follow what it was doing in terms manipulating both `df_participants` and `df_samples` and then modifying them such that they eventually merge into the df. 

Could you maybe explain a bit more about the context and core rationale behind this change? vs the old way of doing merge?

File: docs/source/configuration_file.rst [hunk 2]
```diff
@@ -644,9 +644,9 @@ Split Dataset
         "$schema": "http://json-schema.org/draft-04/schema#",
         "title": "split_method",
         "$$description": [
-            "Metadata contained in a BIDS tabular file on which the files are shuffled, then split\n",
-            "between train/validation/test, according to ``train_fraction`` and ``test_fraction``.\n",
-            "For example, ``participant_id`` from the ``participants.tsv`` file will shuffle all participants,\n",
+            "Metadata contained in a BIDS tabular (TSV) file or a BIDS sidecar JSON file on which the files are shuffled\n",
+            "then split between train/validation/test, according to ``train_fraction`` and ``test_fraction``.\n",
+            "For examples, ``participant_id`` will shuffle all participants from the ``participants.tsv`` file\n",
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply docs/source/configuration_file.rst:649]: It means only the `participant_id` that are present in `participants.tsv`.

If other `sub-XX ` folder are present in the dataset, they won't be used and a warning will now be displayed in the terminal.
In theory, the number of `sub-XX` folder and `participant_id` in a bids dataset SHOULD be the same, but we found that there is no check done in the bids-validator for that.

File: ivadomed/loader/bids_dataframe.py [hunk 3]
```diff
@@ -16,6 +16,8 @@ class BidsDataframe:
         loader_params (dict): Loader parameters, see :doc:`configuration_file` for more details.
         path_output (str): Output folder.
         derivatives (bool): If True, derivatives are indexed.
+        split_method (str): split_method from Split Dataset parameters, see :doc:`configuration_file` for more details.
+            Default: None. Used to remove unused subject files from the bids_dataframe.
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py:20]: Sorry, I don't really understand what you mean, I'll give some details, let me know if that answers the questions.

The `split_method` is a string of the metadata column (any metadata) present in the bids_dataframe. When applying the `train_fraction` and `test_fraction`, `split_method` defines the fraction of _WHAT_ (typically: fraction of `participant_id` of fraction of `sample_id`, but could be fraction of any other metadata).

File: ivadomed/loader/bids_dataframe.py [hunk 4]
```diff
@@ -148,6 +154,16 @@ def create_bids_dataframe(self):
         columns.remove('path')
         self.df = self.df[~(self.df.astype(str).duplicated(subset=columns, keep='first'))]
 
+        # Remove subject files without the "split_method" metadata if specified
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py:157]: The `split_method` is a string of the metadata column (any metadata) present in the bids_dataframe, as detailed in a previous [comment](https://github.com/ivadomed/ivadomed/pull/1112#discussion_r835932428).

File: ivadomed/loader/bids_dataframe.py (hunk 1 above)
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py:216]: I'll refer you to point 1 the first description of the PR [here](https://github.com/ivadomed/ivadomed/pull/1112#issue-1177327116), and try to explain it in other words here as well, sorry I know this is hard to follow and I appreciate your time reviewing it.

I'll take a microscopy example because it contains both subjects (participants) and samples.

In BIDS:
- a "subject" is the `XX` in a sub-XX folder.
- a "participant_id" is the same with the sub- prefix, so `sub-XX`.
- a "sample" is the `YY` in a sample-YY filename part.
- a "sample_id" is the same with the sample- prefix, so `sample-YY`.

You can check the bids_dataframe.csv file to see those 4 columns.

The old way:
1. All the files in the dataset were indexed in the df with 1 row/file --> that create the "subject" and "sample" columns in the main df.
2. The "participant_id" and sample_id" columns were added based on the content of  the "subject" and "sample" columns in the main df.
3. The content of the TSV files were read in `df_participants` and `df_samples`.
4. The content of these df were merged in the main df on "subject" for `participants.tsv` and on "participant_id" and "sample_id" for `samples.tsv.`

Why is it a problem:
At point 2 above: we were assigning the columns "participant_id" and "sample_id" BEFORE reading the TSV file. So there was a strong assumption that all the "participant_id" and "sample_id" were present in the TSV files, which SHOULD be always true but is unfortunately not.

The new way:
1. All the files in the dataset were indexed in the df with 1 row/file --> that create the "subject" and "sample" columns in the main df.
2. Step 2 from above is skipped.
3. The content of the TSV files were read in `df_participants` and `df_samples`.
4. The content of these df are merged in the main df on "subject" for `participants.tsv` and on "subject" and "sample" for `samples.tsv`.

What is the difference?
In the new way, the resulting "participant_id" and "sample_id" columns directly reflects the content of the `participants.tsv` and `samples.stv` files without any assumptions.

File: ivadomed/loader/bids_dataframe.py [hunk 5]
```diff
@@ -25,10 +27,11 @@ class BidsDataframe:
         extensions (list of str): List of file extensions of interest.
         contrast_lst (list of str): List of the contrasts of interest.
         derivatives (bool): If True, derivatives are indexed.
+        split_method (str): split_method from Split Dataset parameters
         df (pd.DataFrame): Dataframe containing dataset information
     """
 
-    def __init__(self, loader_params, path_output, derivatives):
+    def __init__(self, loader_params, path_output, derivatives, split_method=None):
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py]: Done in [66e1f18](https://github.com/ivadomed/ivadomed/pull/1112/commits/66e1f18a4ec3adcf4c62d0deeecb405d9fc92279), for all the methods in the file.

File: ivadomed/loader/bids_dataframe.py [hunk 6]
```diff
@@ -148,6 +154,16 @@ def create_bids_dataframe(self):
         columns.remove('path')
         self.df = self.df[~(self.df.astype(str).duplicated(subset=columns, keep='first'))]
 
+        # Remove subject files without the "split_method" metadata if specified
+        # Keep all derivatives
+        if self.split_method:
+            files_remove = (self.df[(~self.df['path'].str.contains('derivatives')
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py]: Thanks for the advice! Done in [c210092](https://github.com/ivadomed/ivadomed/pull/1112/commits/c210092a4abc2275bb0caac047d6fc79c4b379f6).

File: ivadomed/loader/bids_dataframe.py [hunk 7]
```diff
@@ -148,6 +154,16 @@ def create_bids_dataframe(self):
...
+                               & self.df[self.split_method].isnull())]['filename']).tolist()
+            if files_remove:
+                logger.warning("The files {} don't have the '{}' metadata used as split_method. Skipping files."
+                               .format(files_remove, self.split_method))
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py]: Done in [60eb271](https://github.com/ivadomed/ivadomed/pull/1112/commits/60eb271a4bf8ce207e2eed77b2139a690c05ffdc) for all the logger warnings in the file.

File: ivadomed/loader/bids_dataframe.py [hunk 8]
```diff
@@ -148,6 +154,16 @@ def create_bids_dataframe(self):
...
+                self.df = self.df[~self.df['filename'].str.contains('|'.join(files_remove))]
```
[the remark this answers is not in the document]
Comment (mariehbourget) [reply ivadomed/loader/bids_dataframe.py:170]: Done in [003ec78](https://github.com/ivadomed/ivadomed/pull/1112/commits/003ec78b37b60de59a3b4f1d0cb2ebb8d5283ed1)

File: docs/source/configuration_file.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mariehbourget) [reply docs/source/configuration_file.rst:652]: Yes probably, I'll take a note to follow-up on the BIDS side.
This is not only a "microscopy" issue but concerns both our BEP (for the samples.tsv file) and BIDS in general (for the participants.tsv file)
